ydb/library/yql/providers/yt/provider/yql_yt_provider_context.cpp

```diff
@@ -0,0 +1,214 @@
+            result = val;
+        }
+    }
+    return result;
+}
+
+double ComputeCardinality(TMaybe<double> maxUniques, double leftCardinality, double rightCardinality) {
+    if (!maxUniques) {
+        double result = 0.2 * leftCardinality * rightCardinality;
+        return result;
+    }
+    if (*maxUniques == 0.0) {
+        return 0.0;
+    }
+    double result = leftCardinality / *maxUniques * rightCardinality;
+    return result;
+}
+
+TYtProviderContext::TYtProviderContext(TJoinAlgoLimits limits, TVector<TYtProviderRelInfo> relInfo)
+    : Limits_(limits)
+    , RelInfo_(std::move(relInfo)) {
+    for (int relIndex = 0; relIndex < std::ssize(RelInfo_); ++relIndex) {
+        const auto& rel = RelInfo_[relIndex];
+        for (int colIndex = 0; colIndex < std::ssize(rel.ColumnInfo); ++colIndex) {
+            const auto& column = rel.ColumnInfo[colIndex].ColumnName;
+            ColumnIndex_[column].insert(std::make_pair(relIndex, colIndex));
+        }
+    }
+}
+
+bool TYtProviderContext::IsMapJoinApplicable(const TOptimizerStatistics& stat) const {
+    const TYtProviderStatistic* specific = static_cast<const TYtProviderStatistic*>(stat.Specific.get());
+    return stat.Type == EStatisticsType::BaseTable && specific->SizeInfo.MapJoinMemSize &&
+        *specific->SizeInfo.MapJoinMemSize < Limits_.MapJoinMemLimit;
+}
+
+bool TYtProviderContext::IsLookupJoinApplicable(const TOptimizerStatistics& table, const TOptimizerStatistics& lookupTable, const TVector<TString>& tableJoinKeys) const {
+    const TYtProviderStatistic* tableSpecific = static_cast<const TYtProviderStatistic*>(table.Specific.get());
+    const TYtProviderStatistic* lookupTableSpecific = static_cast<const TYtProviderStatistic*>(lookupTable.Specific.get());
+    if (tableSpecific->SortColumns.empty()) {
+        return false;
+    }
+    if (std::find(tableJoinKeys.begin(), tableJoinKeys.end(), tableSpecific->SortColumns[0]) ==
+        tableJoinKeys.end()) {
+        return false;
+    }
+    if (lookupTable.Nrows > Limits_.LookupJoinMaxRows ||
+        !lookupTableSpecific->SizeInfo.LookupJoinMemSize || *lookupTableSpecific->SizeInfo.LookupJoinMemSize >= Limits_.LookupJoinMemLimit) {
+        return false;
+    }
+
+    return true;
+}
+
+bool TYtProviderContext::IsJoinApplicable(
+    const std::shared_ptr<IBaseOptimizerNode>& left,
+    const std::shared_ptr<IBaseOptimizerNode>& right,
+    const std::set<std::pair<NDq::TJoinColumn, NDq::TJoinColumn>>& /*joinConditions*/,
+    const TVector<TString>& leftJoinKeys,
+    const TVector<TString>& /*rightJoinKeys*/,
+    EJoinAlgoType joinAlgo,
+    EJoinKind /*joinKind*/) {
+    if (joinAlgo == EJoinAlgoType::LookupJoin) {
+        return !leftJoinKeys.empty() && IsLookupJoinApplicable(*left->Stats, *right->Stats, leftJoinKeys);
+    }
+    if (joinAlgo == EJoinAlgoType::MapJoin) {
+        return IsMapJoinApplicable(*right->Stats);
+    }
+    return joinAlgo == EJoinAlgoType::MergeJoin;
```

Comment:
This is somewhat wierd from a clarity point of view
Why not just return true at this point?

Reply:
MergeJoin is used as a generic join algorithm
If we simply return true here, CBO will enumerate many more plans for different values of join algo enum.